### ISSUE TITLE
sending recived audio from contanier to multiple sepakers

### ISSUE THREAD

**Issue by courteous** [caprica/vlcj#1077]
Hello Mark,

here is question regard regarding VLCJ capabilities:
Currently I have a working example of sending and receiving a  MPEG-TS stream container that encapsulates video and audio (H-264 +  MPEG -4 Audio  AAC) which is being send as unicast over udp. Now I found myself in need to "duplicate" the received audio from that container on the receiver side into another speaker of the Laptop, without changing the default speaker. For example I have USB bluetooth headset and laptop speaker. I would like them bought to receive a copy of the received audio stream. 

I have seen that the  embeddedMediaPlayer has a an  AudioApi. And in that API there are couple of methods i.e. 

```
audioAPI.setOutputDevice()
audioAPI.outputDevices()
```

can something like this be done trough dose methods?

The other idea what i can thing of is to do deep to the Java Mixer, but here the question is how to get that audio from the container into formatted byte array  so I can feed it the correct SourceDataLine. Something like this where the "expectedAaudioStream" which should represent a TargetDataLine is the missing part with its bytebuffer data .

```
AudioFormat format = new AudioFormat(8000.0f, 16, 1, true, true);
....
SourceDataLine speakers = null;
speakers = AudioSystem.getSourceDataLine(format, mxInfo);

....
speakers.open(format);
speakers.start();
	 while (bytesRead < Integer.MAX_VALUE) {
		numBytesRead = expectedAaudioStream.read(data, 0, CHUNK_SIZE);
                        bytesRead += numBytesRead;
                        speakers.write(data, 0, numBytesRead);
                    }
speakers.drain();
speakers.close();
expectedAaudioStream.close();
....

```

 Maybe there is already build in into the API that can be used for something like this. I was even thinking about the [duplicate ](https://wiki.videolan.org/Documentation:Streaming_HowTo/Advanced_Streaming_Using_the_Command_Line/#duplicate)method of the VLC, but this can be used only on the stream sending side not on the receiving side. 
can you please point me to the right direction here?

May thanks in advanced!!!

courteous

**Comment by caprica** [comment]
Interesting question.

Setting the output device via the API can only select a single device, so setting a second would unset the first.

You CAN get the audio as a byte[]  and play it via JavaSound, or some other API - have a look here https://github.com/caprica/vlcj-examples/blob/49cb52b76869002fcc4816fb418a8ab80c5e68ac/src/main/java/uk/co/caprica/vlcj/test/directaudio/JavaSoundTest.java

Beware that you are responsible for playing the audio if you do this, so if you want two outputs, it is on YOU to send the audio to both outputs.

Obviously that won't be as efficient as having VLC play it, but for audio only it might be good enough for what you need.

The one caveat is I never really understood why/how to use the presentation timestamp value in the callback - it seems to play audio just fine without it.

And why couldn't you use duplicate to send to your audio to two different outputs? I've never done it myself, but why is that not an option?

**Comment by courteous** [comment]
Hello Mark, 

thank you for the detailed response. I already expected that the 
```
audioAPI.setOutputDevice()
audioAPI.outputDevices()
```
will most likelly not work, but it was worth a shot. That means that I need to concentrate now on
the Java API and work with the TargetDataLine  and SourceDataLine  in order to implement what I what. May thanks for the link. I guess I need to sit down and spend 1 or 2 days in order to test that and I will have the time  this Friday to do that. I will share the code if successful.  

Regarding the duplicate method. Well I thought this will not work since this is the receiving side i.e. it looks like this on the receiving side. I have not find any way do pass string parameters to the VLC on the receving side as we do on the sending side i.e. here is a comparison:

```
this.mediaPlayerFactory = new MediaPlayerFactory(this.myVlcDebugVariable);
this.embeddedMediaPlayer = mediaPlayerFactory.mediaPlayers().newEmbeddedMediaPlayer();
String sourceOfMedia = "udp://@" + ":" + 5555;
MediaApi mediaApi = embeddedMediaPlayer.media();
mediaApi.prepare(sourceOfMedia);
mediaApi.play(sourceOfMedia);
```


in the contrary the sending side I can pass string options i.e. example:



```

  String[] options = { 	
	    						
	    						":dshow-vdev="
	    						+ cameraDevice,  
	    						":dshow-adev="
	    						+ microfonDevice,
	    						":dshow-aspect-ratio=4\\:3",
	    						":live-caching=300",
	    						":sout-x264-tune zerolatency",
	    						":sout-x264-vbv-maxrate " + videoBitrate,
    							":sout=#duplicate{dst=display,"
	    						+ " dst='transcode{vcodec="
	    						+ mediaQualityCodecString
	    						+ ","
	    						+ "fps=" + vlcFpsString
	    						+ "vb=" + videoBitrate
	    						+ ",scale=Auto,"
	    						+ "width=" + vlcPixelWidthString	
	    						+ ",height=" + vlcPixelHeightString
	    						+ ",acodec=mp4a,ab=128,channels=2,samplerate=44100,threads=2,scodec=none} "
	    						+ " :duplicate{ dst=standard{access=udp,mux=ts,dst="
	    						+ ipAddress
	    						+ ":"
	    						+ vlcDestinationTargetStreamingPortInt
	    						+ "}, "
	    						+ ""
	    						+ " dst=standard{access=file,mux=ts,dst="
	    						+ fullFileName
	    						+ "}"
	    						+ " }' "

	    						
	    						
	    						};
	    mediaApi.play(sourceOfMedia, options );


```


So basically the  question here is can I pass VLC string options on the receiving side to the stream, if yes i might be able to do it more easily with VLC directly ?

**Comment by courteous** [comment]
please ignore my last commend i just fond out that mediaApi.play has overloaded methods which can be used with the string options  i.e. that means i can try the duplicate method.

**Comment by courteous** [comment]
i think i need to do what this person is doing here
https://www.thefreewindows.com/19097/vlc-audio-output-sets-speakers-possible/

**Comment by caprica** [comment]
Doesn't that link just show a way to do a quick toggle between one audio output device and the other. It does not send to more than audio output device concurrently.

I don't see how #duplicate can work when receiving a network stream tbh.

The sample I linked to earlier is very easy to comprehend, there's hardly any code to look at.

**Comment by courteous** [comment]
you might me right but, i do not know that yet if duplicate can work or not to be honest , the idea was to see  if i can duplicate the stream and simultaneously use different audio outputs , since that would be the quick solution. If that does not work the only option left here is to use the TargetDataLine and SourceDataLine i.e. the second solution and the link what you have send. Thank very much you for that!!!  I am on it. I will come back to you later this week with some feedback.

**Comment by courteous** [comment]
Hello Mark, 

the second solution works like a charm. As i promised here is the part of the code that worked for me. With this i am able to hear the output on 2 speakers simultaneously.  The one is a USB bluetooth headphone and the second one is the speaker on my laptop . May thanks for the advice and for the help. 


```
package org.utils;


import javax.sound.sampled.AudioFormat;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.Line;
import javax.sound.sampled.LineUnavailableException;
import javax.sound.sampled.Mixer;
import javax.sound.sampled.SourceDataLine;
import javax.sound.sampled.DataLine.Info;

import com.sun.jna.Pointer;

import uk.co.caprica.vlcj.player.base.MediaPlayer;
import uk.co.caprica.vlcj.player.base.callback.AudioCallbackAdapter;


public class JavaSoundCallback  extends AudioCallbackAdapter {

	
	 private static final int BLOCK_SIZE = 4;

	   private static final int SAMPLE_BITS = 16; // BLOCK_SIZE * 8 / channels ???

	   private final AudioFormat audioFormat;

//	   private Info info;

	   private SourceDataLine rptSpeaker;
	   private SourceDataLine fwSpeaker;

	   private Integer rptSpeakerNumber;
	   private Integer fwSpeakerNumber;
	   
	   

	   public JavaSoundCallback(
			   Integer rptSpeakerNumber,
			   Integer fwSpeakerNumber,
			   String format, 
			   int rate, 
			   int channels
			   ) throws Exception {
		   
		   
		   this.rptSpeakerNumber = rptSpeakerNumber;
		   this.fwSpeakerNumber =  fwSpeakerNumber;
		   
	       this.audioFormat = new AudioFormat(rate, SAMPLE_BITS, channels, true, false);
	       
			Mixer.Info[] mixInfors = AudioSystem.getMixerInfo();
			final Mixer.Info rptMxInfo = mixInfors[this.rptSpeakerNumber];
			final Mixer.Info fwMxInfo = mixInfors[this.fwSpeakerNumber];
			


			try {

				this.rptSpeaker = AudioSystem.getSourceDataLine(audioFormat, rptMxInfo);
				this.fwSpeaker = AudioSystem.getSourceDataLine(audioFormat, fwMxInfo);
			} catch (LineUnavailableException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}

			
			
		    start();
	       
	   }

	   private void start() throws Exception {
	       System.out.println("start()");
			rptSpeaker.open(audioFormat);
			rptSpeaker.start();


			fwSpeaker.open(audioFormat);
			fwSpeaker.start();
				       
	   }

	   private void stop() {
	       System.out.println("stop()");
	       rptSpeaker.close();
	       
	       
	       fwSpeaker.close();
	   }

	   @Override
	   public void play(MediaPlayer mediaPlayer, Pointer samples, int sampleCount, long pts) {

	       int bufferSize = sampleCount * BLOCK_SIZE;

	       byte[] data = samples.getByteArray(0, bufferSize);
	       rptSpeaker.write(data, 0, bufferSize);
	       fwSpeaker.write(data, 0, bufferSize);
	       
	       
	       
	       
	       
	   }

	   @Override
	   public void drain(MediaPlayer mediaPlayer) {
	       System.out.println("drain()");
	       rptSpeaker.drain();
	       fwSpeaker.drain();
	   }
	
}



```

**Comment by caprica** [comment]
Excellent news!

What I like about it is you have had to add very little of your own code, vlcj has done almost all the heavy-lifting here.

I might add some note or something in the vlcj JavaSound example as this is I think is very useful information to share.
